lib/src/revset.rs

```diff
@@ -2415,6 +2415,14 @@ pub trait Revset: fmt::Debug {
     /// to how much effort should be put into the estimation, and how accurate
     /// the resulting estimate should be.
     fn count_estimate(&self) -> (usize, Option<usize>);
+
+    /// Returns a revset that caches visited entries.
+    fn caching(&self) -> Box<dyn CachedRevset + '_>;
+}
+
+pub trait CachedRevset {
```

Comment:
I was going to ask if we want this to extend `Revset`, but I can't think of any callers that want both `iter_graph()` and `contains()`, for example. Actually, do we even need `iter()` on `CachedRevset`? I feel like revsets will typically be for either iterating over or for checking membership, but not both. Sorry I didn't think of this before. What do you think, @yuja?

Reply:
> Actually, do we even need `iter()` on `CachedRevset`? I feel like revsets will typically be for either iterating over or for checking membership, but not both.

Good point. I can't think of a good example. In Mercurial, revsets can be composed, so it's sometimes useful to calculate base revset (for iteration), and narrow it down (for membership test.) It's not the case with jj.

There may be GUI use cases to iterate the same revset more than once, but I have no idea.

Another possible future is that the `CachedRevset` is merged in to the `Revset`. If extra `Vec<IndexPosition>` and `RefCell` costs aren't significant, and if we can address the lifetime issue, it will simplify the API.

Reply:
If `CachedRevset` doesn't have an `iter()`, then maybe `.caching()` and `CachedRevset` are not the best names. Maybe `.membership()` and `RevsetMembers` are better? There's probably something better.

Reply:
`.set()`? but meh, `RevsetSet`.

Reply:
Naming is hard :)

I've removed `iter`. I also can't think of any use cases where it might be needed.

Thinking about good name too...

Reply:
FWIW, I've got a use for CachedRevset and do not need iter(). I intend to check whether commits are in the ancestors of immutable_heads() - those commits will be coming from a separate iterator.

Reply:
I was thinking, instead of having `fn caching(&self) -> Box<dyn CachedRevset + '_>;` (or alternatives) in `Revset`, we would have a method that returns a closure, like `fn containing_fn(&self) -> Box<dyn Fn(&CommitId) -> bool + '_>;` (or `contains_fn`). Internally the closure will still use the struct to keep the state but it will be hidden and named however we want. It would allow us to not introduce a new public trait and would tie relatively nicely with what `Revset` does. However, that won't allow us to have an `iter` method alongside in the future.

The usage example would be, for example:
```rust
// ...
let immutable_revset =
    workspace_command.evaluate_revset(workspace_command.immutable_revset()?)?;
let is_commit_immutable = immutable_revset.containing_fn();
// ...
if is_commit_immutable(&commit_id) {
// ...
```

Reply:
That sounds good to me.

Reply:
Sounds good, as long as the boxed closure can be stored and reused. Would it have the same lifetime as the revset from which it was created?

Reply:
> as long as the boxed closure can be stored and reused. Would it have the same lifetime as the revset from which it was created?

The lifetime will be more restricted (borrowing `&self`.) I have a plan to relax it to `'index` lifetime, but it would require non-trivial refactoring.

Reply:
> a plan to relax it to `'index` lifetime, but it would require non-trivial refactoring.

Confirmed that it's doable. I'll follow up. The interface will be something like
```rust
    fn containing_fn<'a>(&self) -> Box<dyn Fn(&CommitId) -> bool + 'a> where Self: 'a;
```

Reply:
I'm a bit worried about the interior mutability problem. Storing both the revset and the containing_fn that can be used to check it is going to run into the usual inability to point to a sibling. Would it be possible for .containing_fn() to consume the revset and take ownership of its data?

Reply:
> I'm a bit worried about the interior mutability problem. Storing both the revset and the containing_fn that can be used to check it is going to run into the usual inability to point to a sibling.

`containing_fn<'a>(&'a self) -> 'a` doesn't allow that (the return value borrows `self`), but `containing_fn<'a>(&self) -> 'a where Self: 'a` should work fine. It can of course be changed to `containing_fn<'a>(Box<Self>) -> 'a where Self: 'a` to eliminate internal data cloning, but that's not necessary?

Reply:
I see, the second signature implies that containing_fn() will extract whatever resources it needs from the revset and not hang onto the reference itself. Looks good.